Add a cheatsheet

DjebbZ opened the pull request:
First draft of a cheatsheet. Far from finished. I put all of Specter API from the Wiki pages (List of navigators and macros).

I used a bit more categories and subcategories than we talked in https://github.com/nathanmarz/specter/issues/56. But the categorization is a draft. And there is a `Don't know yet` category for things I didn't know where to put.

Looking forward to your review!

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+
+###### Transducer-related
+
+`traverse`, `traverse-all`
```
`traverse` is not transducer-related

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+
+###### Multiple elements (return a vector of values found)
+
+`select`, `select-any`
```
`select-any` returns a single value

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+
+#### Sets
+
+`NONE-ELEM`, `set-elem`
```
missing `subset`

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+`NAME`, `NAMESPACE`
+
+#### Atoms
+
```
missing `ATOM`

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+
+#### Views
+
+`NIL->LIST`, `NIL->SET`, `NIL->VECTOR`, `filterer`, `nil->val`, `subselect`, `transform`, `traversed`, `view`
```
`filterer` should go under Sequence

nathanmarz replied:
`transformed`, not `transform`

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+`NIL->LIST`, `NIL->SET`, `NIL->VECTOR`, `filterer`, `nil->val`, `subselect`, `transform`, `traversed`, `view`
+
+#### Value collection
+
```
missing `DISPENSE`

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+
+#### Don't know yet
+
+`NONE`, `codewalker`, `each-nav`, `multi-path`, `must`, `parser`, `walker`
```
`NONE` is not a navigator and should be removed, `eachnav` belongs under "Custom Navigators", `must` belongs under "Maps", `parser` fits best under "Views", `multi-path` belongs under "Control", and `walker`/`codewalker` should get their own section "Walking"

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,120 @@
+
+#### Custom paths
+
+`declarepath`, `defprotocolpath`, `extend-protocolpath`, `path`, `providepath`, `recursive-path`
```
missing `local-declarepath`

nathanmarz commented:
Made inline comments. More generally, I think there's over-categorization here. A cheatsheet should just be a quick reference based on broad criteria. I think this should put all the query operations in one category, all the sequence operations into one category, all the custom navigator operations into one category, etc.

Also, this needs every API element to link to either the API docs or the wiki pages to be useful.

DjebbZ commented:
Fixed almsot all categorization issues. Remains `traverse`, still not sure where it should beloing. In `Sequences` ?

Regarding links, sure it was part of the plan, I just wanted to get categorization right first, because links pollute the reading of the PR. Will do it after your next review.

DjebbZ commented:
Also, I plan to align all function names/links horizontally not vertically, so it looks more like a usual cheatsheet (which is often 2-dimensional).

DjebbZ commented:
Normally it's all good. I put `traverse` in `Query`, added links everywhere and manually verified all of them. I also fixed several links/anchors in the current wiki pages in the process.

Nothing beats a review though.

Quite happy with the current shape of it, even though I just realised there's no mention of the `compiled-*` version of the API in the cheat sheet. I think there should be.

DjebbZ commented:
Now that I added the compiled versions I think we're _really_ all good.

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,105 @@
+
+#### Control
+
+[`STAY`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#stay), [`STOP`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#stop), [`comp-paths`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#comp-paths), [`cond-path`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#cond-path), [`continue-then-stay`](https://github.com/nathanmarz/specter/wiki ...
```
`comp-paths` is not a navigator and should be removed.

DjebbZ replied:
Where should it go then ? Under the new  `Compiled` section in `Operations` ?

DjebbZ replied:
Since you want to remove the `compiled-*` version this one should be removed too, got it.

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,105 @@
+
+[`terminal`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal), [`terminal-val`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal-val)
+
+## Extending Specter
```
These three categories should be merged and named "Custom navigators"

DjebbZ replied:
Ok

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,105 @@
+
+#### Views
+
+[`NIL->LIST`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-list), [`NIL->SET`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-set), [`NIL->VECTOR`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-vector), [`nil->val`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#nil-val), [`parser`](https://github.com/nathanmarz/specter/wi ...
```
`subselect` doesn't belong here. Probably the best place to put it is "Control".

DjebbZ replied:
Seems strange to me since the wiki says it's a view, but, ok...

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,105 @@
+
+#### Transform
+
+[`transform`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#transform), [`multi-transform`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#multi-transform), [`replace-in`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#replace-in), [`setval`](https://github.com/nathanmarz/specter/wiki/List-of-Macros#setval)
```
Should add `vtransform`

DjebbZ replied:
Missed that since it's not in the Wiki.

nathanmarz commented on Cheat-Sheet.md:
```diff
@@ -0,0 +1,105 @@
+
+#### Multi-transform
+
+[`terminal`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal), [`terminal-val`](https://github.com/nathanmarz/specter/wiki/List-of-Navigators#terminal-val)
```
Should add `vterminal`

DjebbZ replied:
Missed that since it's not in the Wiki.

nathanmarz commented:
Looks really good overall. Made more inline comments.

Agree that changing the formatting to be more like a usual cheatsheet would be a big improvement.

I don't think the compiled-* operations belong in the cheatsheet as they are pretty much never needed.

One last navigator missing from the list is `compact`. Probably fits best being listed under "Maps", "Sequences", and "Sets".

DjebbZ commented:
Fixed all the points you mentionned, apart from 

> Agree that changing the formatting to be more like a usual cheatsheet would be a big improvement.

Do you mean that the formatting currently in the PR is the improvement or that it needs to be improved ? I basically put categories one under the other (normal document flow) and put each API item in the same line (horizontally). This is what I meant by "improvement". Are you saying you'd like something else ?

DjebbZ commented:
Realized I forgot to add `local-declarepath`.

DjebbZ commented:
For this PR to be really complete, you @nathanmarz need to add documentation for the API that isn't in the wiki and link to it from the Cheat Sheet. I believe you're the only one able to do it. I looked at the changelog, there are hints as to how to document them but I'm not sure copy-pasting the changelog would be enough.

nathanmarz commented:
This is great, thanks. I also linked to it from the README. I made a note at https://github.com/nathanmarz/specter-wiki/issues/17 to document those last few things.

DjebbZ commented:
You're welcome. By the way I'm curious about how you merged it in the real wiki, since it's a different repository. Did you just set `specter.wiki` as a different remote in git ?

nathanmarz commented:
Yes, that's exactly what I have set up on my local computer.

DjebbZ commented:
Good to know, thanks for the pointer.
